Documentation for match_only_text field (#6041)
opensearch-project/documentation-website: _field-types/supported-field-types/match-only-text.md

---
layout: default
title: Match-only text
nav_order: 61
has_children: false
parent: String field types
grand_parent: Supported field types
---

# Match-only text field type

> A `match_only_text` field is a variant of a text field designed for full-text search when scoring and positional information of terms within a document are not critical.

Reviewer: Should "text" be in code font?

> A `match_only_text` field is a variant of a text field designed for full-text search when scoring and positional information of terms within a document are not critical.

Reviewer 2: ```suggestion
A `match_only_text` field is a variant of a `text` field designed for full-text search when scoring and positional information of terms within a document are not critical.
```

A `match_only_text` field is different from a `text` field in the following ways:

>  - Omits storing positions, frequencies, and norms, reducing storage.

Reviewer: "reducing storage requirements"?

>  - Omits storing positions, frequencies, and norms, reducing storage.

Reviewer 2: ```suggestion
 - Omits storing positions, frequencies, and norms, reducing storage requirements.
```

>  - Disables scoring so all matching documents receive a constant score of 1.0.

Reviewer: ```suggestion
 - Disables scoring so that all matching documents receive a constant score of 1.0.
```

 - Supports all query types except interval and span queries.

> Choose the `match_only_text` field type to prioritize efficient full-text search over complex ranking and positional queries while optimizing storage costs. Using `match_only_text` creates significantly smaller indexes, which translates to lower storage costs, especially for large datasets. 

Reviewer: ```suggestion
Choose the `match_only_text` field type to prioritize efficient full-text search over complex ranking and positional queries while optimizing storage costs. Using `match_only_text` creates significantly smaller indexes, which results in lower storage costs, especially for large datasets. 
```

> Use a `match_only_text` field when you need to quickly find documents containing specific terms without the overhead of storing frequencies and positions. The `match_only_text` field type is not the best choice for ranking results based on relevance or for queries that rely on term proximity or order, like interval or span queries. While this field type does support phrase queries, their performance isn't as efficient as with the `text` field type. If identifying exact phrases or their locations within documents is essential, use the `text` field type instead.

Reviewer: ```suggestion
Use a `match_only_text` field when you need to quickly find documents containing specific terms without the overhead of storing frequencies and positions. The `match_only_text` field type is not the best choice for ranking results based on relevance or for queries that rely on term proximity or order, like interval or span queries. While this field type does support phrase queries, their performance isn't as efficient as when using the `text` field type. If identifying exact phrases or their locations within documents is essential, use the `text` field type instead.
```

## Example

Create a mapping with a `match_only_text` field:

```json
PUT movies
{
  "mappings" : {
    "properties" : {
      "title" : {
        "type" :  "match_only_text"
      }
    }
  }
}
```
{% include copy-curl.html %}

## Parameters

> While `match_only_text` supports most parameters available for text fields, modifying most of them can be counterproductive. This field type thrives on simplicity and efficiency, minimizing data stored in the index to optimize storage costs. Therefore, keeping the default settings is generally the best approach. Any adjustments beyond analyzer settings can reintroduce overhead and negate the efficiency benefits of `match_only_text`.

Reviewer: ```suggestion
While `match_only_text` supports most parameters available for text fields, modifying most of them can be counterproductive. This field type is intended to be simple and efficient, minimizing data stored in the index to optimize storage costs. Therefore, keeping the default settings is generally the best approach. Any modifications beyond analyzer settings can reintroduce overhead and negate the efficiency benefits of `match_only_text`.
```

> While `match_only_text` supports most parameters available for text fields, modifying most of them can be counterproductive. This field type thrives on simplicity and efficiency, minimizing data stored in the index to optimize storage costs. Therefore, keeping the default settings is generally the best approach. Any adjustments beyond analyzer settings can reintroduce overhead and negate the efficiency benefits of `match_only_text`.

Reviewer: Should "text" be in code font?

The following table lists all parameters available for `match_text_only` fields.

Parameter | Description
:--- | :---

> `analyzer` | The analyzer to be used for this field. By default, it will be used at index time and at search time. To override it at search time, set the `search_analyzer` parameter. Default is the `standard` analyzer, which uses grammar-based tokenization and is based on the [Unicode Text Segmentation](https://unicode.org/reports/tr29/) algorithm.

Reviewer: ```suggestion
`analyzer` | The analyzer to be used for the field. By default, it will be used at index time and at search time. To override it at search time, set the `search_analyzer` parameter. Default is the `standard` analyzer, which uses grammar-based tokenization and is based on the [Unicode Text Segmentation](https://unicode.org/reports/tr29/) algorithm.
```

`boost` |  All hits are assigned a score of 1 and are multiplied by `boost` to produce the final score for the query clause.
`eager_global_ordinals` | Specifies whether global ordinals should be loaded eagerly on refresh. If the field is often used for aggregations, this parameter should be set to `true`. Default is `false`.

> `fielddata` | A Boolean value that specifies whether to access analyzed tokens for this field for sorting, aggregation, and scripting. Default is `false`.

Reviewer: ```suggestion
`fielddata` | A Boolean value that specifies whether to access analyzed tokens for sorting, aggregation, and scripting. Default is `false`.
```

> `fielddata_frequency_filter` | A JSON object that specifies to load into memory only those analyzed tokens whose document frequency is between the `min` and `max` values (provided as either an absolute number or a percentage). Frequency is computed per segment. Parameters: `min`, `max`, `min_segment_size`. Default is to load all analyzed tokens.

Reviewer: ```suggestion
`fielddata_frequency_filter` | A JSON object specifying that only those analyzed tokens whose document frequency is between the `min` and `max` values (provided as either an absolute number or a percentage) should be loaded into memory. Frequency is computed per segment. Parameters: `min`, `max`, `min_segment_size`. Default is to load all analyzed tokens.
```

> `fields` | To index the same string in several ways (for example, as a keyword and text), provide the fields parameter. You can specify one version of the field to be used for search and another to be used for sorting and aggregations.

Reviewer: ```suggestion
`fields` | To index the same string in several ways (for example, as a keyword and text), provide the fields parameter. You can specify one version of the field to be used for search and another to be used for sorting and aggregation.
```

> `fields` | To index the same string in several ways (for example, as a keyword and text), provide the fields parameter. You can specify one version of the field to be used for search and another to be used for sorting and aggregations.

Reviewer: Should "fields" be in code font?

`index` | A Boolean value that specifies whether the field should be searchable. Default is `true`.
`index_options` | You cannot modify this parameter.
`index_phrases` | Not supported.
`index_prefixes` | Not supported.
`meta` | Accepts metadata for this field.
`norms` | Norms are disabled and cannot be enabled.

> `position_increment_gap` | Although positions are disabled, `position_increment_gap` behaves similarly to the text field for phrase queries. Such queries may be slower but are still functional.

Reviewer: ```suggestion
`position_increment_gap` | Although positions are disabled, `position_increment_gap` behaves similarly to the text field when used in phrase queries. Such queries may be slower but are still functional.
```

> `position_increment_gap` | Although positions are disabled, `position_increment_gap` behaves similarly to the text field for phrase queries. Such queries may be slower but are still functional.

Reviewer: Should "text" be in code font?

`similarity` | Setting similarity has no impact. The `match_only_text` field type doesn't support queries like `more_like_this`, which rely on similarity. Use a `keyword` or `text` field for queries that rely on similarity.

> `term_vector` | Term vectors are supported but using them is discouraged because it contradicts the primary purpose of this field---storage optimization.

Reviewer: ```suggestion
`term_vector` | Term vectors are supported, but using them is discouraged because it contradicts the primary purpose of this field---storage optimization.
```

## Migrating a field from `text` to `match_only_text`

> You can use the [Reindex API]({{site.url}}{{site.baseurl}}/api-reference/document-apis/reindex/) to migrate from text field to `match_only_text` by updating the correct mapping in the destination index.

Reviewer: ```suggestion
You can use the [Reindex API]({{site.url}}{{site.baseurl}}/api-reference/document-apis/reindex/) to migrate from a text field to `match_only_text` by updating the correct mapping in the destination index.
```

> You can use the [Reindex API]({{site.url}}{{site.baseurl}}/api-reference/document-apis/reindex/) to migrate from text field to `match_only_text` by updating the correct mapping in the destination index.

Reviewer: Should "text" be in code font?

> In the following example, the `source` index has a `title` field of type `text`.

Reviewer: ```suggestion
In the following example, the `source` index contains a `title` field of type `text`.
```

---
layout: default
title: Match only text
nav_order: 61
has_children: false
parent: String field types
grand_parent: Supported field types
redirect_from:
  - /opensearch/supported-field-types/match-only-text/
  - /field-types/match-only-text/
---

# Match only text field type

A variant of text field designed for full-text search when scoring and positional information of terms within a document are not critical.

Key features:
 - Omits positions, frequencies, and norms reducing storage.
 - Disables scoring, resulting in all matching documents receiving a constant score of 1.0.
 - Supports most query types, but not interval/span queries.

> Choosing `match_only_text` means prioritizing efficient full-text search over complex ranking and positional queries, while also optimizing storage costs. It excels when you need to quickly find documents containing specific terms without the overhead of storing frequencies and positions, leading to significantly smaller indexes. This translates to lower storage costs, especially for large datasets. However, it's not the best choice for ranking results based on relevance or for queries that hinge on term proximity or order, like interval/span queries. While it does support phrase queries, their performance isn't as efficient as with text field type. So, if pinpointing exact phrases or their locations within documents is essential, consider using text field type.

Reviewer 3: > While it does support phrase queries, their performance isn't as efficient as with text field type. So, if pinpointing exact phrases or their locations within documents is essential, consider using text field type.

Does this maybe downplay it too much? `match_only_text` supports phrase queries but performance will likely be very very bad, right? I'd maybe tweak the last sentence:

"So, if pinpointing exact phrases or their locations within documents is essential, use the `text` field type instead."

Author: Makes sense! In addition to what you're suggesting, maybe I can add a line on how phrase queries work here and how bad/optimal they could depending on query and workload.

Author: done

## Example

Create a mapping with a text field:

```json
PUT movies
{
  "mappings" : {
    "properties" : {
      "title" : {
        "type" :  "match_only_text"
      }
    }
  }
}
```
{% include copy-curl.html %}

## Parameters

> While `match_only_text` supports most parameters available for text fields, modifying most of them can actually be counterproductive. This field type thrives on simplicity and efficiency, minimizing data stored in the index to optimize storage costs. Therefore, sticking with the default settings is generally the best approach. Any adjustments beyond analyzer settings could reintroduce overhead and negate the efficiency benefits of `match_only_text`.

Reviewer 4: Do the various parameters work? Like, could you theoretically coerce a `match_only_text` field to behave more like a `text` field? (If so, I don't think we should support those parameters. Could we just ignore them, maybe with a deprecation warning?)

Author: All unsupported params or unsupported values for a param will result in an error. But there are other parameters like `stored`, `term_vectors` which are still supported but shouldn't be enabled as it defeats the purpose. 
I like your idea to be more precise here, so I will list down the ones which are supported.

Author: done, thanks

## Migrating from text field

> Reindex API can be used to migrate from text field to `match_only_text` by updating the right mapping in the target index.

Reviewer 4: Maybe add an example of this?

Author: sure
